### PR TITLE
[C#] Add randomness Concept Exercise

### PR DESCRIPTION
**Pull request by mikedamay**
The files in the initial PR were created from V3-dashboard but went astray.  The PR is reconstructed here replacing #1651

### CODE REVIEW

**Comment by mikedamay** [comment]
Revisions have been made in response to [this comment](https://github.com/exercism/v3/issues/1642#issuecomment-645847475)

Please review _design.md_ _Example.cs_ and _instructions.md_.  All other files except _hints.md_ and _csharp/config.json_ are available to show the direction of travel.

**Comment by mikedamay** [comment]
I am comfortabe with this brief story.  I can't imagine students complaining.  However, if we wanted to add heft (and some complexity) then a task to generate unique random numbers would be good and fit well with the spell strength angle.

**Comment by mikedamay** [comment]
test to see the effect of updates on a closed PR

**Comment by mikedamay** [comment]
Please review _Example.cs_, _instructions.md_ and _design.md_.

All other files except _hints.md_ are substantially complete and show the direction of travel.

File: languages/csharp/exercises/concept/randomness/RandomnessTests.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+using Xunit;
+using System;
+
+public class UserDefinedExceptionsTests
+{
+    [Fact]
+    public void RollDie()
+    {
+        var player = new Player();
+        var pips = player.RollDie();
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/randomness/RandomnessTests.cs:10]: One thing we could do here is to call `RollDie` a lot of times (e.g. 100 times or so), to verify that they all are in the right range. Another test that we could add is to call `RollDie` a lot of times and verify that it won't return the same value each time (to make sure students are not using a seed).

File: languages/csharp/exercises/concept/randomness/.meta/design.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+- Know that once generated random numbers can be easily used for many purposes (including non-numeric ones).
+
+## Out of scope
+
+- Secure random numbers
+
+## Concepts
+
+- `randomness`: as they are handled in C# and .NET.
+- `basic-thread-safety`: ensure user is acquainted with the idea of thread safety. A deeper discussion can be deferred to the `concurrent-colletions` exercise.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/randomness/.meta/design.md:15]: Why do we need to introduce thread safety here?

File: languages/csharp/exercises/concept/randomness/.meta/design.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+- Know that both integers and real numbers can be generated
+- Know that once generated random numbers can be easily used for many purposes (including non-numeric ones).
+
+## Out of scope
+
+- Secure random numbers
+
+## Concepts
+
+- `randomness`: as they are handled in C# and .NET.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/randomness/.meta/design.md:14]:
```suggestion
- `randomness`: know how to implement randomness with `System.Random` in C#; know that no seed is required; know that both integers and real numbers can be generated; know that once generated random numbers can be easily used for many purposes (including non-numeric ones)
```

File: languages/csharp/exercises/concept/randomness/.meta/Example.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
+using System;
+
+public class Player
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/randomness/.meta/Example.cs:3]:
```suggestion
public static class Player
```

File: languages/csharp/exercises/concept/randomness/.meta/Example.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+{
+    private static Random random = new Random();
+    private const int DIE_HIGH = 18;
+    public int RollDie()
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/randomness/.meta/Example.cs:7]:
```suggestion
    public static int RollDie()
```

File: languages/csharp/exercises/concept/randomness/.meta/Example.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    {
+        return random.Next(1, DIE_HIGH + 1);
+    }
+
+    public double GenerateSpellStrength()
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/randomness/.meta/Example.cs:12]:
```suggestion
    public static double GenerateSpellStrength()
```

File: languages/csharp/exercises/concept/randomness/.docs/instructions.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+Implement a `RollDie()` method on the `Player` class.
+
+```csharp
+var player = new Player();
+player.RollDie();
+// => >= 1 <= 18
+```
+
+## 2. Players need their strength. Provide a means to generate spell strength
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/randomness/.docs/instructions.md:13]: The two methods (roll die and generate spell strength) seem to relatively unrelated to each other. I'm wondering if we can combine them into a more coherent narrative. 

GenerateSpellStregngth does not seem to make a ton of sense in the domain of D&D, but I also don't have a great idea here. Maybe @SleeplessByte has an idea on how we can solve this?

**Comment by mikedamay** [comment]
temporarily closed while updating

File: languages/csharp/exercises/concept/randomness/.meta/Example.cs (hunk 4 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/.meta/Example.cs:3]: Why static?  If everything else is equal (and I think it is here) then is it not better to get the student out of the habit of thinking in terms of static classes?

File: languages/csharp/exercises/concept/randomness/.meta/design.md (hunk 3 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/.meta/design.md:14]: Mmm.  I have seen that a number of the concepts descriptions follow this pattern.  I feel that this is a description of the exercise not the concept.

What can we say about randomness?  I don't think we want so much to attempt a definition of randomness but perhaps to put it in the context of C# and (as it revolves) round a library API) get the student in the habit of thinking about _.NET_ as a thing.

File: languages/csharp/exercises/concept/randomness/.meta/design.md (hunk 2 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/.meta/design.md:15]: We make the point that a good Random object is a static Random object.  If we are making that point then we should say that it is not thread-safe.  As we haven't introduced thread safety as a concept I  thought we should, particularly as we have to explain it.  I am more than happy to drop it from the concepts in this exercise and just discuss it in the _after.md_.

File: languages/csharp/exercises/concept/randomness/RandomnessTests.cs (hunk 1 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/RandomnessTests.cs:10]: >  verify that they all are in the right range.

I assumed that we would have a few representations which would take care of this.  I am loth to add a test that will occasionally give a false positive (or negative or whichever way it is).  

Are you thinking that we would just make multiple calls in a test?   If we include a hash set in the test file I would like to avoid adding the hash set as a dependency.  As I think I said before I did half the haskell course without understanding the test framework and it really didn't matter.  

>  verify that it won't return the same value each time (to make sure students are not using a seed).

Similar comment I suppose.  I was toying with the idea of getting the student to subclass `Random` and generate `enum`s and return that to the caller, but it doesn't get us much further than creating two players (assuming they are not static, and rolling the die 3 or 4 times.

**Comment by mikedamay** [comment]
I seem to have turned rather argumentative over this short exercise.

File: languages/csharp/exercises/concept/randomness/.meta/Example.cs (hunk 4 above)
Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/randomness/.meta/Example.cs:3]: Well, personally I make everything static _unless_ I can't. But maybe that is just me. There is no state here that is linked to a single player instance so that is why I would prefer static. I'll post a comment in our Slack channel to ask for someone else to chime in.

File: languages/csharp/exercises/concept/randomness/.meta/design.md (hunk 3 above)
Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/randomness/.meta/design.md:14]: > I have seen that a number of the concepts descriptions follow this pattern. I feel that this is a description of the exercise not the concept.

It actually should be sort of a combination: what are the things about the Concept that a student will know after completing the exercise. Does that make more sense?

File: languages/csharp/exercises/concept/randomness/.meta/design.md (hunk 2 above)
Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/randomness/.meta/design.md:15]: > I am more than happy to drop it from the concepts in this exercise and just discuss it in the after.md.

I think I would prefer this. It could actually be in the learning objectives, as that can include the things in the `after.md` document too.

File: languages/csharp/exercises/concept/randomness/RandomnessTests.cs (hunk 1 above)
Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/randomness/RandomnessTests.cs:10]: > I am loth to add a test that will occasionally give a false positive (or negative or whichever way it is).

This could indeed be the case for the uniqueness test, so let's not include that. However, the "is the value in the right range" test should not be prone to false positives (or negatives), or am I missing something?

**Comment by ErikSchierboom** [comment]
> I seem to have turned rather argumentative over this short exercise.

All in the name of getting this exercise right! You've raised good points and I think it is immensely valuable to have these types of discussions upfront, before actually building the entire exercise.

File: languages/csharp/exercises/concept/randomness/.meta/Example.cs (hunk 4 above)
Comment (archrisV) [reply languages/csharp/exercises/concept/randomness/.meta/Example.cs:3]: I think I agree with Erik on this one. As far as I can tell the only purpose of the class is to generate numbers within a certain range without storing them internally. I don't really see the need to create a Player object for that? Am I missing something here?

File: languages/csharp/exercises/concept/randomness/RandomnessTests.cs (hunk 1 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/RandomnessTests.cs:10]: >  am I missing something?

The value we're really interested in is 18 (one below the upper bound) and possibly ensuring that 0 does not turn up.  Representations are going to give us our reassurance there as we can't rely on a test showing this.  Have I missed the point about representations?

File: languages/csharp/exercises/concept/randomness/.meta/design.md (hunk 3 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/.meta/design.md:14]: I suppose.  Concept == chunk of knowledge.  Naming really is very difficult.

File: languages/csharp/exercises/concept/randomness/.docs/instructions.md (hunk 7 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/.docs/instructions.md:13]: I'm happy to put this in an improvement issue if we can't resolve it at present.

File: languages/csharp/exercises/concept/randomness/.meta/Example.cs (hunk 4 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/.meta/Example.cs:3]: Leaving aside he debate about the non-static classes I think there are two downsides to making `Player` a static class in this particular exercise:

- We will no longer be teaching the student to favor a static class for `System.Random` which i think is well worth teaching.

- We end up with rather unrealistic code.  Any game is bound to have multiple players and any player is bound to have state and more capabilities.  I don’t think that because we sacrifice this baggage in the interests of simplicity that we should also sacrifice coding realism.

We're not doing much in this exercise.  We should not miss the opportunity for students to get more familiar with instantiated classes.

File: languages/csharp/exercises/concept/randomness/RandomnessTests.cs (hunk 1 above)
Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/randomness/RandomnessTests.cs:10]: > The value we're really interested in is 18 (one below the upper bound) and possibly ensuring that 0 does not turn up. Representations are going to give us our reassurance there as we can't rely on a test showing this. Have I missed the point about representations?

I get the feeling we're arguing different things :) This test verifies that the output of the `RollDie()` method is in the range 1-18. This is good. As we're dealing with randomness, it could be that the student has set an invalid upper bound. Suppose that upper bound is 19, that would mean that this test fails only 1/19th of a time. Calling the `RollDie()` method 100 times greatly reduces this chance (although it doesn't eliminate it).

> Representations are going to give us our reassurance there as we can't rely on a test showing this

Well yes, but in general anything that we _can_ do to help guide students towards the right solution without it involving an analyzer or representation is a win in my mind.

File: languages/csharp/exercises/concept/randomness/.meta/Example.cs (hunk 4 above)
Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/randomness/.meta/Example.cs:3]: > We will no longer be teaching the student to favor a static class for System.Random which i think is well worth teaching. if everything is static they never have to go through that thought process.

This is a good point.

> We end up with rather unrealistic code. Any game is bound to have multiple players and any player is bound to have state and more capabilities. I don’t think that because we sacrifice this baggage in the interests of simplicity that we should also sacrifice coding realism.

I see what you are saying here. Let's keep it as it is then.

File: languages/csharp/exercises/concept/randomness/.meta/design.md (hunk 2 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/.meta/design.md]: Done

File: languages/csharp/exercises/concept/randomness/RandomnessTests.cs (hunk 1 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/RandomnessTests.cs]: > it could be that the student has set an invalid upper bound. Suppose that upper bound is 19, that would mean that this test fails only 1/19th of a time

We loop 100 times.  The problem is that it does not catch the most likely error.  The upper bound is exclusive so the most common mistake is to set it to 18 (inclusive) rather than 19 which is what it should be.  We cannot easily detect that 18 never turns up. 

It does catch the second most likely error: setting the lower bound to 0 either explicitly or implicitly.

> but in general anything that we can do to help guide students towards the right solution without it involving an analyzer or representation is a win in my mind.

I agree with the sentiment but I would understand  this better if I knew the workflow.  If a student is operating in the browser do they run all the tests  by pressing a button or something after they have completed all tasks?

File: languages/csharp/exercises/concept/randomness/RandomnessTests.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        for (int i = 0; i < 100; i++)
+        {
+            var pips = player.RollDie();
+            if (pips <= 0 || pips > 18)
+            {
+                outOfRange = true;
+                break;
+            }
+        }
+        Assert.False(outOfRange);
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/randomness/RandomnessTests.cs:20]: Was there any reason for not doing:

```suggestion
        var player = new Player();
        for (var i = 0; i < 100; i++)
        {
            Assert.InRange(player.RollDie(), 1, 18);
        }
```

File: languages/csharp/exercises/concept/randomness/.docs/instructions.md (hunk 7 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/.docs/instructions.md:13]: Raised improvement issue #1749

Comment (SleeplessByte) [reply languages/csharp/exercises/concept/randomness/.docs/instructions.md:13]: Sorry missed this! Feel free to ping on slack when I'm not responsive. I'm pretty good in managing my time :)

Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/randomness/.docs/instructions.md:13]: :)

Comment (mikedamay) [reply languages/csharp/exercises/concept/randomness/.docs/instructions.md:13]: @SleeplessByte There is actually a story-improvement issue, #1749.  We would appreciate any input. The C# exercises may be forked by others.  At least that was @ErikSchierboom 's hope before I got involved.
